// ngx-image-zoom/esm2020/lib/zoom-modes/hover-freeze-zoom-mode.mjs
export class HoverFreezeZoomMode {
    constructor(zoomService) {
        this.zoomService = zoomService;
        this.zoomFrozen = false;
    }
    onClick(event) {
        if (this.zoomService.zoomingEnabled && this.zoomFrozen) {
            this.zoomFrozen = false;
        }
        else if (this.zoomService.zoomingEnabled) {
            this.zoomFrozen = true;
            this.zoomService.markForCheck();
        }
        else {
            this.zoomService.zoomOn(event);
        }
    }
    onMouseEnter(event) {
        if (!this.zoomFrozen) {
            this.zoomService.zoomOn(event);
        }
    }
    onMouseLeave() {
        if (this.zoomService.zoomingEnabled && !this.zoomFrozen) {
            this.zoomService.zoomOff();
        }
    }
    onMouseMove(event) {
        if (this.zoomService.zoomingEnabled && !this.zoomFrozen) {
            this.zoomService.calculateZoomPosition(event);
        }
    }
    onMouseWheel() {
        // Prevent scroll zoom if we're frozen
        return !this.zoomFrozen;
    }
}
